java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java

```diff
@@ -496,8 +496,13 @@ private Resolution selectBest(Env env, JavaType defSite, JavaType callSite, List
     }
 
     Resolution resolution = new Resolution(mostSpecific);
-    JavaType returnType = ((MethodJavaType) ((JavaSymbol.MethodJavaSymbol) mostSpecific).type).resultType;
-    resolution.type = typeSubstitutionSolver.getReturnType(returnType, defSite, callSite, !typeParams.isEmpty(), substitution);
+    JavaSymbol.MethodJavaSymbol mostSpecificMethod = (JavaSymbol.MethodJavaSymbol) mostSpecific;
+    if(substitution.isComplete(mostSpecificMethod)) {
```

Comment:
Testing for completeness is probably too strong and implies that the _defered_ type will be used too often. By the way, _defer**r**ed_ is the correct spelling. 

Example:

```
class A {
  abstract <T, U> T foo(T t);
  void bar() {
    foo("hello"); // return type is deferred when we already knows it's a String
  }
}
```